File: Data Structures & Algorithms/duplicate-integer/submission-1.py
# Time Complexity: O(n), single pass through list
# Space Complexity: O(n), set stores up to n unique elements

class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        hashset = set()
        for i in nums:
            if i not in hashset:
                hashset.add(i)
            else: 
                return True
        return False

File: Data Structures & Algorithms/remove-duplicates-from-sorted-array/submission-1.java
// Time complexity: O(n)
// Space complexity: Two pointers, O(1)

class Solution {
    public int removeDuplicates(int[] nums) {
        int n = nums.length;
        int l = 0, r = 0;
        while (r < n) {
            nums[l] = nums[r];
            // Shift right ptr until new unique num is found
            while (r < n && nums[r] == nums[l]) {
                r++;
            } 
            // Shift left ptr to store new found unique num
            l++;
        }
        // Leaves anything from index k onward unspecified
        
        return l;
    }
}